streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorTest.java

```diff
@@ -0,0 +1,203 @@
+import org.apache.kafka.streams.processor.MockProcessorContext;
+import org.apache.kafka.test.MockInternalProcessorContext;
+import org.hamcrest.BaseMatcher;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.junit.Test;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Random;
+
+import static java.time.Duration.ZERO;
+import static org.apache.kafka.streams.kstream.Suppressed.BufferConfig.unbounded;
+import static org.apache.kafka.streams.kstream.Suppressed.IntermediateSuppression.withEmitAfter;
+import static org.apache.kafka.streams.kstream.Suppressed.emitFinalResultsOnly;
+import static org.apache.kafka.streams.kstream.Suppressed.intermediateEvents;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.fail;
+
+@SuppressWarnings("PointlessArithmeticExpression")
+public class KTableSuppressProcessorTest {
+    // Using a static seed gives us deterministic (aka repeatable) tests while documenting which test data are arbitrary
+    private static final Random RANDOM = new Random(42);
+
+    private static long arbitraryLong() {
+        return RANDOM.nextLong();
+    }
+
+    private static TimeWindow arbitraryWindow() {
+        final int end = RANDOM.nextInt(Integer.MAX_VALUE);
+        return new TimeWindow(RANDOM.nextInt(end), end);
+    }
+
+    @Test
+    public void zeroEmitAfterShouldImmediatelyEmit() {
+        final KTableSuppressProcessor<String, Long> processor =
+            new KTableSuppressProcessor<>(getImpl(intermediateEvents(withEmitAfter(ZERO))));
+
+        final MockInternalProcessorContext context = new MockInternalProcessorContext();
+        processor.init(context);
+
+        final long timestamp = arbitraryLong();
+        context.setTimestamp(timestamp);
+        context.setStreamTime(timestamp);
+        final String key = "hey";
+        final Change<Long> value = new Change<>(arbitraryLong(), arbitraryLong());
+        processor.process(key, value);
+
+        assertThat(context.forwarded(), hasSize(1));
+        final MockProcessorContext.CapturedForward capturedForward = context.forwarded().get(0);
+        assertThat(capturedForward.keyValue(), is(new KeyValue<>(key, value)));
+        assertThat(capturedForward.timestamp(), is(timestamp));
+    }
+
+    @Test
+    public void windowedZeroEmitAfterShouldImmediatelyEmit() {
+        final KTableSuppressProcessor<Windowed<String>, Long> processor =
+            new KTableSuppressProcessor<>(getImpl(intermediateEvents(withEmitAfter(ZERO))));
+
+        final MockInternalProcessorContext context = new MockInternalProcessorContext();
+        processor.init(context);
+
+        final long timestamp = arbitraryLong();
+        context.setTimestamp(timestamp);
+        context.setStreamTime(timestamp);
+        final Windowed<String> key = new Windowed<>("hey", arbitraryWindow());
```

Comment:
Should `timestamp` not fall into the window boundaries? Not sure if it's important for the test?

Reply:
In practice, it would, but since the processor itself doesn't make any assumption between the record timestamp and the window boundary, it doesn't matter for the test. In fact, the test verifies that the processor makes no such assumption.